In [2]:
import pandas as pd

#load data
path= r'C:\Users\prett\OneDrive\Desktop\CODING\PYTHON\ANALYSIS\Portfolio\portfolio.xlsx'
df = pd.read_excel(path)

#display data
df


Country Name Country Code  \
0                   OECD members          OED   
1                   OECD members          OED   
2                   OECD members          OED   
3                   OECD members          OED   
4                   OECD members          OED   
...                          ...          ...   
2075  Africa Western and Central          AFW   
2076  Africa Western and Central          AFW   
2077  Africa Western and Central          AFW   
2078  Africa Western and Central          AFW   
2079  Africa Western and Central          AFW   

                                            Series Name     Series Code  \
0                             Deposit interest rate (%)     FR.INR.DPST   
1                             Lending interest rate (%)     FR.INR.LEND   
2               Pump price for gasoline (US$ per liter)  EP.PMP.SGAS.CD   
3            Pump price for diesel fuel (US$ per liter)  EP.PMP.DESL.CD   
4            Multidimensional poverty index (scale 0-1)  SI.POV.MDIM.XQ   
...                                                 ...             ...   
2075            Access to electricity (% of population)  EG.ELC.ACCS.ZS   
2076  Cause of death, by communicable diseases and m...  SH.DTH.COMM.ZS   
2077  Maternal mortality ratio (modeled estimate, pe...     SH.STA.MMRT   
2078                          Number of maternal deaths     SH.MMR.DTHS   
2079                          Number of neonatal deaths     SH.DTH.NMRT   

     1973 [YR1973] 1974 [YR1974] 1975 [YR1975] 1976 [YR1976] 1977 [YR1977]  \
0               ..            ..            ..            ..            ..   
1               ..            ..            ..            ..            ..   
2               ..            ..            ..            ..            ..   
3               ..            ..            ..            ..            ..   
4               ..            ..            ..            ..            ..   
...            ...           ...           ...           ...           ...   
2075            ..            ..            ..            ..            ..   
2076            ..            ..            ..            ..            ..   
2077            ..            ..            ..            ..            ..   
2078            ..            ..            ..            ..            ..   
2079            ..            ..            ..            ..            ..   

     1978 [YR1978]  ... 2013 [YR2013] 2014 [YR2014] 2015 [YR2015]  \
0               ..  ...            ..            ..            ..   
1               ..  ...            ..            ..            ..   
2               ..  ...            ..         1.725            ..   
3               ..  ...            ..         1.605            ..   
4               ..  ...            ..            ..            ..   
...            ...  ...           ...           ...           ...   
2075            ..  ...      47.05199     47.647405     46.758948   
2076            ..  ...            ..            ..     61.883166   
2077            ..  ...           808           811           788   
2078            ..  ...        125000        127000        125000   
2079            ..  ...        534778        537757        540701   

     2016 [YR2016] 2017 [YR2017] 2018 [YR2018] 2019 [YR2019] 2020 [YR2020]  \
0               ..            ..            ..            ..            ..   
1               ..            ..            ..            ..            ..   
2             1.32            ..            ..            ..            ..   
3             1.16            ..            ..            ..            ..   
4               ..            ..            ..            ..            ..   
...            ...           ...           ...           ...           ...   
2075     50.903108      48.78764     51.212863     51.160817     51.734388   
2076            ..            ..            ..      59.50382            ..   
2077           787           780           779           764  

# Data Preprocessing

In [3]:
# Extract the year portion from column names and truncate to the first four characters
new_columns = [str(year)[:4] for year in df.columns[4:]]

# Update the column names in df by combining the first four columns and new_columns
df.columns = df.columns[:4].tolist() + new_columns


# Display the modified dataframe with updated column names
df



Country Name Country Code  \
0                   OECD members          OED   
1                   OECD members          OED   
2                   OECD members          OED   
3                   OECD members          OED   
4                   OECD members          OED   
...                          ...          ...   
2075  Africa Western and Central          AFW   
2076  Africa Western and Central          AFW   
2077  Africa Western and Central          AFW   
2078  Africa Western and Central          AFW   
2079  Africa Western and Central          AFW   

                                            Series Name     Series Code 1973  \
0                             Deposit interest rate (%)     FR.INR.DPST   ..   
1                             Lending interest rate (%)     FR.INR.LEND   ..   
2               Pump price for gasoline (US$ per liter)  EP.PMP.SGAS.CD   ..   
3            Pump price for diesel fuel (US$ per liter)  EP.PMP.DESL.CD   ..   
4            Multidimensional poverty index (scale 0-1)  SI.POV.MDIM.XQ   ..   
...                                                 ...             ...  ...   
2075            Access to electricity (% of population)  EG.ELC.ACCS.ZS   ..   
2076  Cause of death, by communicable diseases and m...  SH.DTH.COMM.ZS   ..   
2077  Maternal mortality ratio (modeled estimate, pe...     SH.STA.MMRT   ..   
2078                          Number of maternal deaths     SH.MMR.DTHS   ..   
2079                          Number of neonatal deaths     SH.DTH.NMRT   ..   

     1974 1975 1976 1977 1978  ...      2013       2014       2015       2016  \
0      ..   ..   ..   ..   ..  ...        ..         ..         ..         ..   
1      ..   ..   ..   ..   ..  ...        ..         ..         ..         ..   
2      ..   ..   ..   ..   ..  ...        ..      1.725         ..       1.32   
3      ..   ..   ..   ..   ..  ...        ..      1.605         ..       1.16   
4      ..   ..   ..   ..   ..  ...        ..         ..         ..         ..   
...   ...  ...  ...  ...  ...  ...       ...        ...        ...        ...   
2075   ..   ..   ..   ..   ..  ...  47.05199  47.647405  46.758948  50.903108   
2076   ..   ..   ..   ..   ..  ...        ..         ..  61.883166         ..   
2077   ..   ..   ..   ..   ..  ...       808        811        788        787   
2078   ..   ..   ..   ..   ..  ...    125000     127000     125000     127000   
2079   ..   ..   ..   ..   ..  ...    534778     537757     540701     543246   

          2017       2018       2019       2020       2021 2022  
0           ..         ..         ..         ..         ..   ..  
1           ..         ..         ..         ..         ..   ..  
2           ..         ..         ..         ..         ..   ..  
3           ..         ..         ..         ..         ..   ..  
4           ..         ..         ..         ..         ..   ..  
...        ...        ...        ...        ...        ...  ...  
2075  48.78764  51.212863  51.160817  51.734388  54.224187   ..  
2076        ..         ..   59.50382         ..         ..   ..  
2077       780        779        764        733         ..   ..  
2078    127000     129000     129000     125000         ..   ..  
2079    544464     545693     546112     546107     545443   ..  

[2080 rows x 54 columns]

In [4]:
# Melt the DataFrame to make it easier for analysis
df_melted = df.melt(
    id_vars=['Country Name', 'Country Code', 'Series Name', 'Series Code'],  # Columns to keep as identifiers
    var_name='Year',  # Name of the new column for the melted variable names
    value_name='Value'  # Name of the new column for the melted values
)

# Display the dataframee
df_melted


Country Name Country Code  \
0                     OECD members          OED   
1                     OECD members          OED   
2                     OECD members          OED   
3                     OECD members          OED   
4                     OECD members          OED   
...                            ...          ...   
103995  Africa Western and Central          AFW   
103996  Africa Western and Central          AFW   
103997  Africa Western and Central          AFW   
103998  Africa Western and Central          AFW   
103999  Africa Western and Central          AFW   

                                              Series Name     Series Code  \
0                               Deposit interest rate (%)     FR.INR.DPST   
1                               Lending interest rate (%)     FR.INR.LEND   
2                 Pump price for gasoline (US$ per liter)  EP.PMP.SGAS.CD   
3              Pump price for diesel fuel (US$ per liter)  EP.PMP.DESL.CD   
4              Multidimensional poverty index (scale 0-1)  SI.POV.MDIM.XQ   
...                                                   ...             ...   
103995            Access to electricity (% of population)  EG.ELC.ACCS.ZS   
103996  Cause of death, by communicable diseases and m...  SH.DTH.COMM.ZS   
103997  Maternal mortality ratio (modeled estimate, pe...     SH.STA.MMRT   
103998                          Number of maternal deaths     SH.MMR.DTHS   
103999                          Number of neonatal deaths     SH.DTH.NMRT   

        Year Value  
0       1973    ..  
1       1973    ..  
2       1973    ..  
3       1973    ..  
4       1973    ..  
...      ...   ...  
103995  2022    ..  
103996  2022    ..  
103997  2022    ..  
103998  2022    ..  
103999  2022    ..  

[104000 rows x 6 columns]

In [5]:
# Convert 'Year' column to datetime format
df_melted['Year'] = pd.to_datetime(df_melted['Year']).dt.year

# Verify the data type has been changed
print(df_melted['Year'].dtype)


int64


In [6]:
# Drop the specified columns to concentrate on more relevant columns
drop_columns = df_melted.drop(['Country Code', 'Series Code'], axis=1)
# Display the resulting DataFrame
drop_columns.head(5)

Country Name                                 Series Name  Year Value
0  OECD members                   Deposit interest rate (%)  1973    ..
1  OECD members                   Lending interest rate (%)  1973    ..
2  OECD members     Pump price for gasoline (US$ per liter)  1973    ..
3  OECD members  Pump price for diesel fuel (US$ per liter)  1973    ..
4  OECD members  Multidimensional poverty index (scale 0-1)  1973    ..

In [7]:

# Pivot the DataFrame to convert 'Series Name' values to columns
df_pivoted = df_melted.pivot(
    index=['Country Name', 'Year'],
    columns='Series Name',
    values='Value'
)

# Reset the index of the pivoted DataFrame
df_pivoted = df_pivoted.reset_index()

# Display the pivoted DataFrame
df_pivoted.head(5)


Series Name                Country Name  Year  \
0            Africa Western and Central  1973   
1            Africa Western and Central  1974   
2            Africa Western and Central  1975   
3            Africa Western and Central  1976   
4            Africa Western and Central  1977   

Series Name Access to electricity (% of population)  \
0                                                ..   
1                                                ..   
2                                                ..   
3                                                ..   
4                                                ..   

Series Name Cause of death, by communicable diseases and maternal, prenatal and nutrition conditions (% of total)  \
0                                                           ..                                                      
1                                                           ..                                                      
2                                                           ..                                                      
3                                                           ..                                                      
4                                                           ..                                                      

Series Name Consumer price index (2010 = 100) Deposit interest rate (%)  \
0                                          ..                        ..   
1                                          ..                        ..   
2                                          ..                        ..   
3                                          ..                        ..   
4                                          ..                        ..   

Series Name Educational attainment, Doctoral or equivalent, population 25+, total (%) (cumulative)  \
0                                                           ..                                       
1                                                           ..                                       
2                                                           ..                                       
3                                                           ..                                       
4                                                           ..                                       

Series Name Educational attainment, at least Bachelor's or equivalent, population 25+, total (%) (cumulative)  \
0                                                           ..                                                  
1                                                           ..                                                  
2                                                           ..                                                  
3                                                           ..                                                  
4                                                           ..                                                  

Series Name Educational attainment, at least Master's or equivalent, population 25+, total (%) (cumulative)  \
0                                                           ..                                                
1                                                           ..                                                
2                                                           ..                                                
3                                                           ..                                                
4                                                           ..                                                

Series Name Educational attainment, at least completed lower secondary, population 25+, total (%) (cumulative)  \
0                                                           ..                                                   
1                                

In [8]:
# Remove the column title for the index labels ('Series Name')
df_pivoted.columns.name = None

# Display the DataFrame without the column title
df_pivoted.head(5)



Country Name  Year Access to electricity (% of population)  \
0  Africa Western and Central  1973                                      ..   
1  Africa Western and Central  1974                                      ..   
2  Africa Western and Central  1975                                      ..   
3  Africa Western and Central  1976                                      ..   
4  Africa Western and Central  1977                                      ..   

  Cause of death, by communicable diseases and maternal, prenatal and nutrition conditions (% of total)  \
0                                                 ..                                                      
1                                                 ..                                                      
2                                                 ..                                                      
3                                                 ..                                                      
4                                                 ..                                                      

  Consumer price index (2010 = 100) Deposit interest rate (%)  \
0                                ..                        ..   
1                                ..                        ..   
2                                ..                        ..   
3                                ..                        ..   
4                                ..                        ..   

  Educational attainment, Doctoral or equivalent, population 25+, total (%) (cumulative)  \
0                                                 ..                                       
1                                                 ..                                       
2                                                 ..                                       
3                                                 ..                                       
4                                                 ..                                       

  Educational attainment, at least Bachelor's or equivalent, population 25+, total (%) (cumulative)  \
0                                                 ..                                                  
1                                                 ..                                                  
2                                                 ..                                                  
3                                                 ..                                                  
4                                                 ..                                                  

  Educational attainment, at least Master's or equivalent, population 25+, total (%) (cumulative)  \
0                                                 ..                                                
1                                                 ..                                                
2                                                 ..                                                
3                                                 ..                                                
4                                                 ..                                                

  Educational attainment, at least completed lower secondary, population 25+, total (%) (cumulative)  \
0                                                 ..                                                   
1                                                 ..                                                   
2                                                 ..                                                   
3                                                 ..                                                   
4                                                 ..                                                   

   ... Pump price for diesel fuel (US$ per liter)  \
0  ...                                         ..  

In [9]:
# Group the dataframe by Country Name for country-specific analysis
grouped_df = df_pivoted.groupby('Country Name')

# Retrieve the group names (unique values from the 'Country Name' column)
group_names = grouped_df.groups.keys()

# Iterate over each group name and print it
for name in group_names:
    print(name)


Africa Western and Central
Australia
Canada
China
Euro area
Europe & Central Asia
European Union
Germany
Ghana
India
Japan
Kuwait
Latin America & Caribbean
Middle East & North Africa
Nigeria
North America
OECD members
Russian Federation
Saudi Arabia
Seychelles
South Africa
South Asia
Sub-Saharan Africa
United Arab Emirates
United Kingdom
United States


In [10]:
#retrieve group for analysis

group_name = 'United Kingdom'  
group_uk = grouped_df.get_group(group_name)

# show all columns for easy analysis
pd.set_option('display.max_columns', None)

group_uk.head(5)


Country Name  Year Access to electricity (% of population)  \
1200  United Kingdom  1973                                      ..   
1201  United Kingdom  1974                                      ..   
1202  United Kingdom  1975                                      ..   
1203  United Kingdom  1976                                      ..   
1204  United Kingdom  1977                                      ..   

     Cause of death, by communicable diseases and maternal, prenatal and nutrition conditions (% of total)  \
1200                                                 ..                                                      
1201                                                 ..                                                      
1202                                                 ..                                                      
1203                                                 ..                                                      
1204                                                 ..                                                      

     Consumer price index (2010 = 100) Deposit interest rate (%)  \
1200                          11.94469                  8.020833   
1201                         13.861098                       9.5   
1202                         17.216493                  7.083333   
1203                         20.067463                  7.541667   
1204                         23.246202                  4.895833   

     Educational attainment, Doctoral or equivalent, population 25+, total (%) (cumulative)  \
1200                                                 ..                                       
1201                                                 ..                                       
1202                                                 ..                                       
1203                                                 ..                                       
1204                                                 ..                                       

     Educational attainment, at least Bachelor's or equivalent, population 25+, total (%) (cumulative)  \
1200                                                 ..                                                  
1201                                                 ..                                                  
1202                                                 ..                                                  
1203                                                 ..                                                  
1204                                                 ..                                                  

     Educational attainment, at least Master's or equivalent, population 25+, total (%) (cumulative)  \
1200                                                 ..                                                
1201                                                 ..                                                
1202                                                 ..                                                
1203                                                 ..                                                
1204                                                 ..                                                

     Educational attainment, at least completed lower secondary, population 25+, total (%) (cumulative)  \
1200                                                 ..                                                   
1201                                                 ..                                                   
1202                                                 ..                                                   
1203                                                 ..                                                   
1204                                                 ..                                                   

     Educational attainment, at least com

In [11]:
# Due to limitations in the availability of data spanning 50 years, the following columns will be selected from the available statistics.
selected_columns = ['Country Name',
                    'Year',
                    'Consumer price index (2010 = 100)',
                    'GDP (constant 2015 US$)',
                    'GDP growth (annual %)',
                    'GDP per capita (constant 2015 US$)',
                    'GNI growth (annual %)',
                    'General government final consumption expenditure (% of GDP)',
                    'General government final consumption expenditure (constant 2015 US$)',
                    'Government expenditure on education, total (% of GDP)',
                    'Imports of goods, services and primary income (BoP, current US$)',
                    'Income share held by fourth 20%',
                    'Income share held by highest 10%',
                    'Income share held by highest 20%',
                    'Income share held by lowest 10%',
                    'Income share held by lowest 20%',
                    'Income share held by second 20%',
                    'Income share held by third 20%',
                    'Lending interest rate (%)',
                    'Life expectancy at birth, total (years)',
                    'Number of neonatal deaths',
                    'Number of under-five deaths',
                    'Poverty headcount ratio at $6.85 a day (2017 PPP) (% of population)',
                    'Proportion of people living below 50 percent of median income (%)',
                    'Subsidies and other transfers (% of expense)',
                    'Tax revenue (% of GDP)']

# The COVID-affected period (2020-2022), which may introduce confounding or interference, will be excluded.
# The analysis will concentrate on the data from 1973 to 2019.

excluded_df = group_uk[((group_uk['Year'] < 2020) | (group_uk['Year'] > 2022))]
analysis_df = excluded_df[(excluded_df['Year'] >= 1973) & (excluded_df['Year'] <= 2019)]
selected_uk_df = analysis_df[selected_columns]

# Display the new dataframe
selected_uk_df.head(55)



Country Name  Year Consumer price index (2010 = 100)  \
1200  United Kingdom  1973                          11.94469   
1201  United Kingdom  1974                         13.861098   
1202  United Kingdom  1975                         17.216493   
1203  United Kingdom  1976                         20.067463   
1204  United Kingdom  1977                         23.246202   
1205  United Kingdom  1978                         25.167069   
1206  United Kingdom  1979                         28.544811   
1207  United Kingdom  1980                         33.673151   
1208  United Kingdom  1981                         37.672385   
1209  United Kingdom  1982                         40.911782   
1210  United Kingdom  1983                          42.79753   
1211  United Kingdom  1984                         44.920592   
1212  United Kingdom  1985                         47.647898   
1213  United Kingdom  1986                         49.281082   
1214  United Kingdom  1987                         51.325716   
1215  United Kingdom  1988                          53.45848   
1216  United Kingdom  1989                         56.537821   
1217  United Kingdom  1990                         61.096726   
1218  United Kingdom  1991                         65.655632   
1219  United Kingdom  1992                         68.670242   
1220  United Kingdom  1993                         70.427224   
1221  United Kingdom  1994                         71.990013   
1222  United Kingdom  1995                          73.93194   
1223  United Kingdom  1996                         76.040318   
1224  United Kingdom  1997                         77.714074   
1225  United Kingdom  1998                         79.128907   
1226  United Kingdom  1999                         80.515998   
1227  United Kingdom  2000                         81.468467   
1228  United Kingdom  2001                         82.716849   
1229  United Kingdom  2002                         83.974478   
1230  United Kingdom  2003                         85.130387   
1231  United Kingdom  2004                         86.314037   
1232  United Kingdom  2005                         88.117255   
1233  United Kingdom  2006                         90.281117   
1234  United Kingdom  2007                         92.435731   
1235  United Kingdom  2008                         95.690771   
1236  United Kingdom  2009                         97.567967   
1237  United Kingdom  2010                               100   
1238  United Kingdom  2011                        103.856112   
1239  United Kingdom  2012                        106.528574   
1240  United Kingdom  2013                        108.969854   
1241  United Kingdom  2014                        110.551137   
1242  United Kingdom  2015                        110.958017   
1243  United Kingdom  2016                        112.076937   
1244  United Kingdom  2017                        114.943592   
1245  United Kingdom  2018                        117.579064   
1246  United Kingdom  2019                        119.622711   

     GDP (constant 2015 US$) GDP growth (annual %)  \
1200    1262757606009.822754              6.523848   
1201    1231385605650.851074             -2.484404   
1202    1213239297438.027344              -1.47365   
1203    1248547789733.673096              2.910266   
1204    1279233979840.417725              2.457751   
1205    1333016308392.240234               4.20426   
1206    1382991310974.798828              3.749017   
1207     1354897671049.24292             -2.031368   
1208    1344224548669.106689             -0.787744   
1209    1371040366003.380615              1.994891   
1210    1428923720513.524902              4.221856   
1211    1461347502638.866455              2.269105   
1212    1521955648714.161865              4.147415   
1213    1569902434705.617676              3.150341   
1214    1654563164607.023682              5.392738   
1215    1749409568712.282227              5.732414   
1216    179450239

In [12]:
#There are missing values in the 'Lending interest rate (%)' column. Values for Year 2015 - 2019 are missing
# Similarly values for 1978, 1979,1993 and 1997 are missing for 'Government expenditure on education, total (% of GDP)' column
#To allow for smooth operation in a way that does not impact the data negatively will be deviced to fill it up 

# Import libraries
from sklearn.impute import SimpleImputer

# Create a copy of the DataFrame to avoid modifying the original data
selected_uk_df_copy = selected_uk_df.copy()

# Convert '..' to NaN as NaN values are typically treated as missing values in pandas and scikit-learn

for column in ['Lending interest rate (%)', 'Government expenditure on education, total (% of GDP)']:
    selected_uk_df_copy[column] = selected_uk_df_copy[column].replace('..', float('NaN'))

# Perform imputation to fill in missing values in both columns
imputer = SimpleImputer(strategy='mean')
selected_uk_df_copy[['Lending interest rate (%)', 'Government expenditure on education, total (% of GDP)']] = imputer.fit_transform(selected_uk_df_copy[['Lending interest rate (%)', 'Government expenditure on education, total (% of GDP)']])

# Display the updated DataFrame
selected_uk_df_copy.head(10)



Country Name  Year Consumer price index (2010 = 100)  \
1200  United Kingdom  1973                          11.94469   
1201  United Kingdom  1974                         13.861098   
1202  United Kingdom  1975                         17.216493   
1203  United Kingdom  1976                         20.067463   
1204  United Kingdom  1977                         23.246202   
1205  United Kingdom  1978                         25.167069   
1206  United Kingdom  1979                         28.544811   
1207  United Kingdom  1980                         33.673151   
1208  United Kingdom  1981                         37.672385   
1209  United Kingdom  1982                         40.911782   

     GDP (constant 2015 US$) GDP growth (annual %)  \
1200    1262757606009.822754              6.523848   
1201    1231385605650.851074             -2.484404   
1202    1213239297438.027344              -1.47365   
1203    1248547789733.673096              2.910266   
1204    1279233979840.417725              2.457751   
1205    1333016308392.240234               4.20426   
1206    1382991310974.798828              3.749017   
1207     1354897671049.24292             -2.031368   
1208    1344224548669.106689             -0.787744   
1209    1371040366003.380615              1.994891   

     GDP per capita (constant 2015 US$) GNI growth (annual %)  \
1200                        22471.18489              6.759535   
1201                       21899.096124             -2.495104   
1202                       21577.981948             -1.775763   
1203                       22211.422837              2.976298   
1204                       22764.806641              1.941189   
1205                        23720.62697              4.330518   
1206                       24587.844966               3.53675   
1207                       24059.602837             -2.288942   
1208                       23861.764281             -0.733292   
1209                       24346.505423              1.892586   

     General government final consumption expenditure (% of GDP)  \
1200                                             18.271            
1201                                          20.162522            
1202                                          22.441843            
1203                                          22.183646            
1204                                          20.790834            
1205                                          20.279237            
1206                                          19.873144            
1207                                           21.31376            
1208                                          21.755479            
1209                                          21.625081            

     General government final consumption expenditure (constant 2015 US$)  \
1200                                289433772133.422119                     
1201                                295221264556.690186                     
1202                                312046373513.181458                     
1203                                317664594276.089294                     
1204                                313217842913.981201                     
1205                                318854872148.410522                     
1206                                 324010953900.75824                     
1207                                329062248434.787659                     
1208                                329795737574.127197                     
1209                                332184378318.468506                     

      Government expenditure on education, total (% of GDP)  \
1200                                           5.448610       
1201                                           5.391000       
1202                                           6.268780       
1203                                           6.424220       
1204                                           6.063350       
1205             

In [13]:
# the data was imported as object. As part of  getting it ready for statistical analysis, it will be converted to float.
# Get all column names except 'Country Name' and 'Year'
cols_to_convert = [col for col in selected_uk_df_copy.columns if col not in ['Country Name', 'Year']]

# Convert the selected columns to float
for col in cols_to_convert:
    selected_uk_df_copy[col] = selected_uk_df_copy[col].astype(float)



# Exploratory Data Analysis (EDA) and Visualisation

In [14]:
pip install plotly


Note: you may need to restart the kernel to use updated packages.


In [15]:
# To understanding the data get more information. Exclude the 'Country Name' and 'Year' column

selected_uk_df_copy.iloc[:, 2:].describe()


Consumer price index (2010 = 100)  GDP (constant 2015 US$)  \
count                          47.000000             4.700000e+01   
mean                           70.472945             2.088202e+12   
std                            31.211501             6.179838e+11   
min                            11.944690             1.213239e+12   
25%                            46.284245             1.491652e+12   
50%                            76.040318             1.996070e+12   
75%                            94.063251             2.658249e+12   
max                           119.622711             3.174164e+12   

       GDP growth (annual %)  GDP per capita (constant 2015 US$)  \
count              47.000000                           47.000000   
mean                2.139359                        34745.577244   
std                 2.111049                         8516.761652   
min                -4.510486                        21577.981948   
25%                 1.654665                        26406.802031   
50%                 2.430073                        34316.219715   
75%                 3.153814                        42663.328818   
max                 6.523848                        47491.594563   

       GNI growth (annual %)  \
count              47.000000   
mean                2.137217   
std                 2.205770   
min                -4.065061   
25%                 1.184476   
50%                 2.382347   
75%                 3.491856   
max                 6.759535   

       General government final consumption expenditure (% of GDP)  \
count                                          47.000000             
mean                                           19.510607             
std                                             1.638306             
min                                            16.070620             
25%                                            18.593514             
50%                                            19.656798             
75%                                            20.596535             
max                                            22.441843             

       General government final consumption expenditure (constant 2015 US$)  \
count                                       4.700000e+01                      
mean                                        4.244105e+11                      
std                                         1.001055e+11                      
min                                         2.894338e+11                      
25%                                         3.416209e+11                      
50%                                         3.786357e+11                      
75%                                         5.317884e+11                      
max                                         6.053435e+11                      

       Government expenditure on education, total (% of GDP)  \
count                                          47.000000       
mean                                            4.959497       
std                                             0.590565       
min                                             3.989180       
25%                                             4.487465       
50%                                             4.959497       
75%                                             5.380160       
max                                             6.424220       

       Imports of goods, services and primary income (BoP, current US$)  \
count                                       4.700000e+01                  
mean                                        6.171216e+11                  
std                                         4.477972e+11                  
min                                         5.520904e+10                  
25%                                         1.939769e+11                  
50%                                         4.989088e+11                  
75%                                   

In [46]:
import plotly.express as px

# Get all numeric column names except for the first two
numeric_columns = selected_uk_df_copy.columns[2:]

# Create different types of plots for each column
for idx, col in enumerate(numeric_columns[:9]):  # Only include the first 10 columns
    if idx == 0:
        # Scatter plot with color by 'Year'
        fig = px.scatter(selected_uk_df_copy, x='Year', y=col, size=col, title=f'Scatter Plot of {col} Over Years',
                         color='Year', color_continuous_scale=px.colors.sequential.Viridis,
                         template='plotly_dark')
    elif idx == 1:
        # Box plot with pink color
        fig = px.box(selected_uk_df_copy, x='Year', y=col, title=f'Box Plot of {col} Over Years',
                     color_discrete_sequence=['pink'], template='plotly_dark')
    elif idx == 2:
        # Violin plot
        fig = px.violin(selected_uk_df_copy, x='Year', y=col, title=f'Violin Plot of {col} Over Years',
                        color_discrete_sequence=['yellow'], template='plotly_dark')
    elif idx == 3:
        # Area plot
        fig = px.area(selected_uk_df_copy, x='Year', y=col, title=f'Area Plot of {col} Over Years',
                      color_discrete_sequence=px.colors.qualitative.Plotly, template='plotly_dark')
    elif idx == 4:
        # 3D Scatter plot with color by 'Year'
        fig = px.scatter_3d(selected_uk_df_copy, x='Year', y=col, z=col, title=f'3D Scatter Plot of {col} Over Years',
                            color='Year', color_continuous_scale=px.colors.sequential.Viridis,
                            template='plotly_dark')
    elif idx == 5:
        # Polar Bar plot
        fig = px.bar_polar(selected_uk_df_copy, r=col, theta='Year', title=f'Polar Bar Plot of {col} Over Years',
                           color_discrete_sequence=px.colors.sequential.Viridis,  # Use the "Viridis" color scale
                           template='plotly_dark')  # Use the 'plotly_dark' template with black paper color
        fig.update_layout(polar=dict(bgcolor='green'))  # Set green background color for the polar plot
    elif idx == 6:
        # Polar Scatter plot
        fig = px.scatter_polar(selected_uk_df_copy, r=col, theta='Year', title=f'Polar Scatter Plot of {col} Over Years',
                               color_discrete_sequence=px.colors.qualitative.Plotly, template='plotly_dark')
    elif idx == 7:
        # Polar Line plot
        fig = px.line_polar(selected_uk_df_copy, r=col, theta='Year', title=f'Polar Line Plot of {col} Over Years',
                            color_discrete_sequence=px.colors.qualitative.Plotly, template='plotly_dark')
    elif idx == 8:
        # Ternary Scatter plot with color by 'Year'
        fig = px.scatter_ternary(selected_uk_df_copy, a=col, b='Year', c='Year',
                                 title=f'Ternary Scatter Plot of {col} Over Years',
                                 color='Year', color_continuous_scale=px.colors.sequential.Viridis,
                                 template='plotly_dark')

    fig.show()



# Machine Learning

In [47]:
# Normalise data

from sklearn.preprocessing import MinMaxScaler
import pandas as pd

# Create a copy of the original data frame
selected_uk_df_copy_normalized = selected_uk_df_copy.copy()

# Specify the columns to normalize
columns_to_normalize = ['Consumer price index (2010 = 100)', 'GDP (constant 2015 US$)', 'GDP growth (annual %)', 'GDP per capita (constant 2015 US$)', 'GNI growth (annual %)', 'General government final consumption expenditure (% of GDP)', 'General government final consumption expenditure (constant 2015 US$)', 'Government expenditure on education, total (% of GDP)', 'Imports of goods, services and primary income (BoP, current US$)', 'Income share held by fourth 20%', 'Income share held by highest 10%', 'Income share held by highest 20%', 'Income share held by lowest 10%', 'Income share held by lowest 20%', 'Income share held by second 20%', 'Income share held by third 20%', 'Lending interest rate (%)', 'Life expectancy at birth, total (years)', 'Number of neonatal deaths', 'Number of under-five deaths', 'Poverty headcount ratio at $6.85 a day (2017 PPP) (% of population)', 'Proportion of people living below 50 percent of median income (%)', 'Subsidies and other transfers (% of expense)', 'Tax revenue (% of GDP)']

# Apply Min-Max normalization to the specified columns
scaler = MinMaxScaler()
selected_uk_df_copy_normalized[columns_to_normalize] = scaler.fit_transform(selected_uk_df_copy_normalized[columns_to_normalize])

# Print the normalized data frame
selected_uk_df_copy_normalized


Country Name  Year  Consumer price index (2010 = 100)  \
1200  United Kingdom  1973                           0.000000   
1201  United Kingdom  1974                           0.017798   
1202  United Kingdom  1975                           0.048959   
1203  United Kingdom  1976                           0.075436   
1204  United Kingdom  1977                           0.104957   
1205  United Kingdom  1978                           0.122796   
1206  United Kingdom  1979                           0.154164   
1207  United Kingdom  1980                           0.201791   
1208  United Kingdom  1981                           0.238932   
1209  United Kingdom  1982                           0.269016   
1210  United Kingdom  1983                           0.286529   
1211  United Kingdom  1984                           0.306245   
1212  United Kingdom  1985                           0.331574   
1213  United Kingdom  1986                           0.346741   
1214  United Kingdom  1987                           0.365729   
1215  United Kingdom  1988                           0.385536   
1216  United Kingdom  1989                           0.414134   
1217  United Kingdom  1990                           0.456472   
1218  United Kingdom  1991                           0.498811   
1219  United Kingdom  1992                           0.526807   
1220  United Kingdom  1993                           0.543124   
1221  United Kingdom  1994                           0.557638   
1222  United Kingdom  1995                           0.575672   
1223  United Kingdom  1996                           0.595253   
1224  United Kingdom  1997                           0.610797   
1225  United Kingdom  1998                           0.623936   
1226  United Kingdom  1999                           0.636818   
1227  United Kingdom  2000                           0.645664   
1228  United Kingdom  2001                           0.657257   
1229  United Kingdom  2002                           0.668937   
1230  United Kingdom  2003                           0.679672   
1231  United Kingdom  2004                           0.690664   
1232  United Kingdom  2005                           0.707411   
1233  United Kingdom  2006                           0.727506   
1234  United Kingdom  2007                           0.747516   
1235  United Kingdom  2008                           0.777745   
1236  United Kingdom  2009                           0.795179   
1237  United Kingdom  2010                           0.817765   
1238  United Kingdom  2011                           0.853576   
1239  United Kingdom  2012                           0.878395   
1240  United Kingdom  2013                           0.901067   
1241  United Kingdom  2014                           0.915753   
1242  United Kingdom  2015                           0.919531   
1243  United Kingdom  2016                           0.929923   
1244  United Kingdom  2017                           0.956545   
1245  United Kingdom  2018                           0.981021   
1246  United Kingdom  2019                           1.000000   

      GDP (constant 2015 US$)  GDP growth (annual %)  \
1200                 0.025253               1.000000   
1201                 0.009254               0.183616   
1202                 0.000000               0.275217   
1203                 0.018006               0.672515   
1204                 0.033655               0.631505   
1205                 0.061082               0.789785   
1206                 0.086567               0.748527   
1207                 0.072241               0.224673   
1208                 0.066798               0.337378   
1209                 0.080473               0.589558   
1210                 0.109991               0.791379   
1211                 0.126526               0.614409   
1212                 0.157434               0.784633   
1213                 0.181885               0.694272   
1214                 0.225059               0.8

#### Exporting Normalized Data to Excel for Dashboard Building

In [24]:
# Save the normalized data to Excel for dashboard creation
selected_uk_df_copy_normalized.to_excel('uk_normalised_data.xlsx', index=False)

In [48]:
# Exclude 'Country Name' and 'Year' columns from the correlation matrix
columns_to_exclude = ['Country Name', 'Year']
correlation_matrix = selected_uk_df_copy.drop(columns=columns_to_exclude).corr()

# Display the correlation matrix as a table
correlation_table = correlation_matrix.style.background_gradient(cmap='coolwarm')
correlation_table.format("{:.2f}")  # Optional: Format the correlation values to two decimal places
correlation_table



### Model development

In [28]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from joblib import dump, load
import pandas as pd

# Specify the feature columns
X_columns = ['Tax revenue (% of GDP)',
             'Life expectancy at birth, total (years)',
             'Income share held by lowest 20%',
             'GDP growth (annual %)',
             'Consumer price index (2010 = 100)']

# Specify the target column
y_column = 'Proportion of people living below 50 percent of median income (%)'

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(selected_uk_df_copy[X_columns], selected_uk_df_copy[y_column], test_size=0.2, random_state=42)

# Instantiate the model
model = LinearRegression()

# Fit the model
model.fit(X_train, y_train)

# Save the model to a file
dump(model, 'model.joblib')


# Make predictions
y_pred = model.predict(X_test)

# Print the coefficients
coeff_df = pd.DataFrame(model.coef_, X_columns, columns=['Coefficient'])
print(coeff_df)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print(f"Mean Squared Error (MSE): {mse}")
print(f"R-squared (R2): {r2}")


                                         Coefficient
Tax revenue (% of GDP)                      0.033002
Life expectancy at birth, total (years)    -0.595198
Income share held by lowest 20%            -2.184181
GDP growth (annual %)                       0.004674
Consumer price index (2010 = 100)           0.072702
Mean Squared Error (MSE): 0.21543169736114245
R-squared (R2): 0.9553426136769257


##### Model Evaluation 
An MSE of 0.2154 suggests that the model's predictions have relatively low average squared differences from the actual values. This indicates that the model is capturing a significant portion of the variance in the target variable and is performing reasonably well.

A high R2 value of 0.9553 suggests that the model explains a significant proportion of the variance in the target variable.It implies that about 95.4% of the variation in the proportion of people living below 50 percent of median income is explained by the predictors in the model. A high R2 value usually indicates that the model fits the data well